### ISSUE TITLE
Redundant type-checking in the generated JS

### ISSUE THREAD

**Issue by nikita-volkov** [elm-lang/elm-compiler#1065]
Consider the following Elm function:

```haskell    
linear : (Float, Float) -> (Float, Float) -> (Float -> Float)
linear (inputMin, inputMax) (outputMin, outputMax) input =
  outputMin + (input - inputMin) / inputMax * outputMax
```

and the associated compiled JavaScript:

```javascript
var linear = F3(function (_v0,
_v1,
input) {
   return function () {
      switch (_v1.ctor)
      {case "_Tuple2":
         return function () {
              switch (_v0.ctor)
              {case "_Tuple2":
                 return _v1._0 + (input - _v0._0) / _v0._1 * _v1._1;}
              _U.badCase($moduleName,
              "on line 12, column 3 to 56");
           }();}
      _U.badCase($moduleName,
      "on line 12, column 3 to 56");
   }();
});
```

In the generated JS you can see two checks of the constructor name performed and two closures introduced to achieve that. It's quite an overhead I must say! So why waste resources on those checks when compiler's type-checker makes them redundant?

**Comment by jvoigtlaender** [comment]
The same argument can be made for user-defined types with exactly one constructor (but possibly several fields in that constructor):
```elm
type D a b = C a Int b
```
Any `case-of` on a value of some type of the form `D T1 T2` could skip checking the constructor to be `C`.

**Comment by laszlopandy** [comment]
This has already been fixed in elm-compiler master. Evan implemented a
decision tree analyser, which ignores checks which are unnecessary and
generates minimal code to check the rest.

It will be available in Elm 0.16.

**Comment by evancz** [comment]
Yeah, whether it is user defined stuff or tuples or whatever, it is all left out in 0.16 if it is not necessary.
